base/initdefs.jl

```diff
@@ -374,22 +375,48 @@ exit code `n` (instead of the original exit code). If more than one exit hook
 calls `exit(n)`, then Julia will exit with the exit code corresponding to the
 last called exit hook that calls `exit(n)`. (Because exit hooks are called in
 LIFO order, "last called" is equivalent to "first registered".)
+
+Note: Once all exit hooks have been called, no more exit hooks can be registered,
+and any call to `atexit(f)` after all hooks have completed will throw an exception.
+This situation may occur if you are registering exit hooks from background Tasks that
+may still be executing concurrently during shutdown.
 """
-atexit(f::Function) = Base.@lock _atexit_hooks_lock (pushfirst!(atexit_hooks, f); nothing)
+function atexit(f::Function)
+    Base.@lock _atexit_hooks_lock begin
+        _atexit_hooks_finished[] && error("cannot register new atexit hook; already exiting.")
+        pushfirst!(atexit_hooks, f)
+        return nothing
+    end
+end
 
 function _atexit(exitcode::Cint)
-    while !isempty(atexit_hooks)
-        f = popfirst!(atexit_hooks)
-        try
-            if hasmethod(f, (Cint,))
-                f(exitcode)
-            else
-                f()
+    # Don't hold the lock around the iteration, just in case any other thread executing in
+    # parallel tries to register a new atexit hook while this is running. We don't want to
+    # block that thread from proceeding, and we can allow it to register its hook which we
+    # will immediately run here.
+    while Base.@lock _atexit_hooks_lock !isempty(atexit_hooks)
+        f = Base.@lock _atexit_hooks_lock begin
+            popfirst!(atexit_hooks)
+        end
+            try
+                if hasmethod(f, (Cint,))
+                    f(exitcode)
+                else
+                    f()
+                end
+            catch ex
+                showerror(stderr, ex)
+                Base.show_backtrace(stderr, catch_backtrace())
+                println(stderr)
             end
-        catch ex
-            showerror(stderr, ex)
-            Base.show_backtrace(stderr, catch_backtrace())
-            println(stderr)
+
+        Base.@lock _atexit_hooks_lock begin
+            # If this is the last iteration, atomically disable atexit hooks to prevent
+            # someone from registering a hook that will never be run.
+            # (We do this inside the loop, so that it is atomic: no one can have registered
+            #  a hook that never gets run, and we run all the hooks we know about until
+            #  the vector is empty.)
+            isempty(atexit_hooks) && (_atexit_hooks_finished[] = true)
```

Comment:
There is a bit of a race here, because you check `isempty(atexit_hooks)` in two separate places. It is probably cleaner to change to `while true`, then get the lock and do the atomic operations and `break` out of the lock and loop if it is empty

Reply:
Thanks, yeah, that sounds nicer. 👍 I'll make that change now.

------

@kpamnany: also suggested that we could just drop the support for registering atexit hooks from inside other atexit hooks, and just disable registering new hooks as soon as we start processing them.

Does that sound reasonable to you? I agree with him that it would dramatically simplify the code and might also be a clearer API for users? "Once we are exiting, you can no longer register new atexit hooks."

Reply:
Done. But can you also comment on the proposed behavior simplification?

Reply:
@vtjnash ping on this thread, since you might have missed it

Reply:
I don't think that really helps simplify anything, and is minorly breaking

Reply:
Okay cool thanks. I think I agree. 👍 

This should be good to go then! Please take another look?